### PR TITLE
Add `USE_MAGMA` build flag

### PR DESCRIPTION
**Pull request by eqy**
Many model pipelines/workflows don't use MAGMA even though it is included in the build by default. Leaving MAGMA kernels out of the build can save 60+MB of GPU memory when loading `libtorch_cuda.so` (tested on V100, current upstream master).

A current sharp corner of this flag is that toggling it when rebuilding requires `torch/include/THC/THCGeneral.h` to be *manually* deleted by the user, as even running `make clean` or `setup.py` with `--cmake` does not properly regenerate it with the appropriate substitution for `#cmakedefine USE_MAGMA`. Is there a way to force the regeneration of the header during a rebuild?

CC @malfet @ptrblck

### CODE REVIEW

File: cmake/Dependencies.cmake [hunk 1]
```diff
@@ -1604,6 +1606,7 @@ if(NOT INTERN_BUILD_MOBILE)
     message(STATUS "MAGMA V2 check: ${MAGMA_V2}")
   else()
     message(STATUS "MAGMA not found. Compiling without MAGMA support")
+    unset(USE_MAGMA)
```
Comment (walterddr) [inline cmake/Dependencies.cmake:1609]: Is it possible to use:
```suggestion
elseif(USE_MAGMA)
    message(WARNING
      "Not compiling with MAGMA. Suppress this warning with "
      "-DUSE_MAGMA=OFF.")
    caffe2_update_option(USE_MAGMA OFF)
else()
    message(STATUS "MAGMA not found. Compiling without MAGMA support")
endif()
```
Example for this is from the 
```
# ---[ CUDA
``` 
section of the same file

**Comment by walterddr** [comment]
Thanks for the contribution @eqy . This sounds really promising.

Do you have more details regarding the 60+MB reduction on loading `libtorch_cuda.so`, e.g. was it built from source or from pip/conda? if from source do you know what's included in the cuda arch list?

Regarding your second question, do you have a minimum repro and expected behavior / actual behavior comparison so we can verify?

**Comment by eqy** [comment]
> Thanks for the contribution @eqy . This sounds really promising.
> 
> Do you have more details regarding the 60+MB reduction on loading `libtorch_cuda.so`, e.g. was it built from source or from pip/conda? if from source do you know what's included in the cuda arch list?
> 
> Regarding your second question, do you have a minimum repro and expected behavior / actual behavior comparison so we can verify?

Sure, I built pytorch from upstream master with `TORCH_CUDA_ARCH_LIST="7.0" CUDA_HOME="/usr/local/cuda"     CMAKE_PREFIX_PATH="$(dirname $(which conda))/../"     NCCL_INCLUDE_DIR="/usr/include/"     NCCL_LIB_DIR="/usr/lib/"     USE_SYSTEM_NCCL=1 USE_MAGMA=0/1 python setup.py develop`
on the [20.11 NVIDIA container](https://docs.nvidia.com/deeplearning/frameworks/pytorch-release-notes/rel_20-11.html#rel_20-11) with V100 (probably unnecessary for repo). Note that as mentioned before, `torch/include/THC/THCGeneral.h` needs to be manually deleted before toggling `USE_MAGMA` or else the build will fail at link time due to incorrect `#define`s.

To check the memory usage, I simply run 
```
import torch
a = torch.tensor([1.0]).cuda().reshape([1,1,1,1])
```
and look at the current GPU memory used (e.g., with `nvidia-smi`).
It varies by the exact GPU setup or driver, but on V100 I generally see ~860MB used without MAGMA and ~930MB used with MAGMA.

**Review by malfet — APPROVED** [review]
Please rebase your PR against latest `viable/strict` to get good CI signal.

**Pull request closed without merging**